leakpro/synthetic_data_attacks/anonymeter/evaluators/inference_evaluator.py

```diff
@@ -0,0 +1,220 @@
+import pandas as pd
+from pydantic import BaseModel, ConfigDict
+
+from leakpro.import_helper import Self
+from leakpro.synthetic_data_attacks.anonymeter.neighbors.mixed_types_n_neighbors import mixed_type_n_neighbors
+from leakpro.synthetic_data_attacks.anonymeter.stats.confidence import EvaluationResults
+from leakpro.synthetic_data_attacks.anonymeter.utils import assert_x_in_bound
+
+
+class InferenceGuesses(BaseModel):
+    """Utility class to store guesses, secrets and count from inference attack.
+
+    Parameters
+    ----------
+    guesses : pd.Series
+        Attacker guesses for each of the targets.
+    secrets : pd.Series
+        Array with the true values of the secret for each of the targets.
+    regression : bool
+        Whether or not the attacker is trying to solve a classification or
+        a regression task. The first case is suitable for categorical or
+        discrete secrets, the second for numerical continuous ones.
+    tolerance : float, default is 0.05
+        Maximum value for the relative difference between target and secret
+        for the inference to be considered correct.
+
+    """
+
+    model_config = ConfigDict(arbitrary_types_allowed = True)
+    guesses: npt.NDArray
+    secrets: npt.NDArray
+    regression: bool
+    tolerance: float = 0.05
+    matches: Optional[npt.NDArray] = None
+    count: Optional[int] = None
+
+    def __init__(self: Self, **kwargs: npt.NDArray) -> None:
+        super().__init__(**kwargs)
+        #Assert input values
+        assert len(self.guesses.shape) == 1
+        assert self.guesses.shape[0] > 0
+        assert self.guesses.shape == self.secrets.shape
+        assert_x_in_bound(
+            x = self.tolerance,
+            x_name = "tolerance",
+            inclusive_flag = True
+        )
+
+    def evaluate_inference_guesses(self: Self) -> Self:
+        """Evaluate the success of an inference attack.
+
+        The attack is successful if the attacker managed to make a correct guess.
+
+        In case of regression problems, when the secret is a continuous variable,
+        the guess is correct if the relative difference between guess and target
+        is smaller than a given tolerance. In the case of categorical target
+        variables, the inference is correct if the secrets are guessed exactly.
+
+        Returns
+        -------
+        Self : InferenceGuesses
+            Object containing inference guesses and successes (count).
+
+        """
+        #Set guesses and secrets
+        guesses = self.guesses
+        secrets = self.secrets
+        #Calculate value matches (depending on self.regression)
+        if self.regression:
+            rel_abs_diff = np.abs(guesses - secrets) / (guesses + 1e-12)
+            value_matches = rel_abs_diff <= self.tolerance
+        else:
+            value_matches = guesses == secrets
+        #Calculate nan matches
+        nan_matches = np.logical_and(pd.isna(guesses), pd.isna(secrets))
```

Comment:
Is there any reason why we even consider nans in this attack? Is it interesting?